File: day5/s1.py
import sys

rules = {}
lines = []


def check(arr, n):
    for x in range(n):
        for y in range(x+1, n):
            if arr[x] not in rules or arr[y] not in rules[arr[x]]:
                return False
    return True


with open(sys.argv[1]) as file:
    lines = file.readlines()

m = len(lines)
i = 0
while i < m:
    line = lines[i].rstrip()
    if len(line) == 0:
        break
    u, v = map(int, line.split('|'))
    if u in rules:
        rules[u].add(v)
    else:
        rules[u] = set([v])

    i += 1

i += 1
ans = 0
while i < m:
    line = lines[i].rstrip()
    arr = list(map(int, line.split(',')))
    n = len(arr)

    if check(arr, n):
        ans += arr[n // 2]

    i += 1

print(ans)

File: day5/s1.cpp
#include <fstream>
#include <ios>
#include <iostream>
#include <ostream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

std::unordered_map<int, std::unordered_set<int>> rules;
int main(int argc, char* argv[]) {
    std::ifstream in(argv[1], std::ios_base::in);
    std::string line;
    int u, v;
    char c;
    // Rules
    while (std::getline(in, line)) {
        if (line.size() == 0)
            break;
        std::istringstream(line) >> u >> c >> v;
        rules[u].insert(v);
    }

    long long ans = 0;
    // Queries
    while(std::getline(in, line)) {
        std::vector<int> arr;
        std::istringstream iss(line);
        int q;
        while (iss >> q >> c) arr.push_back(q);
        arr.push_back(q);
        int n = arr.size();
        bool poss = true;
        for (int i = 0; poss && i < n; i++)
            for (int j = i+1; poss && j < n; j++)
                if (!rules.count(arr[i]) || !rules[arr[i]].count(arr[j]))
                    poss = false;
        if (poss)
            ans += arr[n / 2];
    }

    std::cout << ans << std::endl;
    return 0;
}
